// node-red-contrib-flujo-estandar/CalcularFlujoEstandar.js
const csv = require('csv-parser');
const fs = require('fs');

/**
 * Calculador de Flujo Estándar para LPG (Butano/Propano)
 * Implementa correcciones según:
 * - API MPMS 11.2.4 (Tabla 23E/24E) - Corrección por temperatura (Ctl)
 * - API MPMS 11.2.2M - Corrección por presión/compresibilidad (Cpl)
 *
 * Basado en tablas oficiales de Measurement Canada
 */
class CalcularFlujoEstandar {
    constructor(csvPath = '/data/api23e.csv') {
        console.log("Iniciando CalcularFlujoEstandar...");

        this.api23Data = [];
        this.dataLoaded = false;

        // Densidad del agua a 60°F (15.56°C) en kg/m³
        this.WATER_DENSITY_60F = 999.016;
        // Densidad del agua a 15°C en kg/m³
        this.WATER_DENSITY_15C = 999.1;

        // Propiedades de referencia para LPG
        this.FLUID_PROPERTIES = {
            'butano': {
                standardDensity: 584,      // kg/m³ a 15°C
                molecularWeight: 58.12
            },
            'propano': {
                standardDensity: 507,      // kg/m³ a 15°C
                molecularWeight: 44.10
            },
            'lpg': {
                standardDensity: 538,      // Mezcla típica
                molecularWeight: 49.71
            }
        };

        // Tablas de presión de vapor basadas en API/Measurement Canada
        // Temperatura en °C, Presión en kPa absoluta
        // AJUSTADAS según comparación con computador de flujo real
        this.vaporPressureTables = {
            // Propano a diferentes densidades (kg/m³ @ 15°C)
            '500': { // Propano ligero
                '-10': 420, '-5': 490, '0': 570, '5': 660, '10': 760,
                '15': 870, '20': 990, '25': 1120, '30': 1260, '35': 1420, '40': 1590
            },
            '505': { // Propano típico
                '-10': 360, '-5': 420, '0': 490, '5': 570, '10': 660,
                '15': 760, '20': 870, '25': 990, '30': 1120, '35': 1270, '40': 1430
            },
            '510': { // Propano pesado
                '-10': 310, '-5': 365, '0': 430, '5': 500, '10': 580,
                '15': 670, '20': 770, '25': 880, '30': 1000, '35': 1140, '40': 1290
            },
            // Butano (aproximado ~580-590 kg/m³)
            '580': {
                '-10': 70, '-5': 90, '0': 115, '5': 145, '10': 180,
                '15': 220, '20': 270, '25': 325, '30': 390, '35': 460, '40': 540
            }
        };

        // Tabla de factores Cpl basada en API 11.2.2M para LPG a 505 kg/m³
        // Formato: [ΔP en kPa][rango temperatura °C] = Cpl
        // Ajustada para coincidir con computadores de flujo comerciales
        this.cplTable = {
            50:   { '-10': 1.00016, '0': 1.00019, '5': 1.00021, '10': 1.00023, '15': 1.00026, '20': 1.00029, '25': 1.00033, '30': 1.00037, '35': 1.00042 },
            100:  { '-10': 1.00032, '0': 1.00038, '5': 1.00042, '10': 1.00046, '15': 1.00052, '20': 1.00058, '25': 1.00065, '30': 1.00074, '35': 1.00084 },
            150:  { '-10': 1.00048, '0': 1.00057, '5': 1.00063, '10': 1.00069, '15': 1.00078, '20': 1.00087, '25': 1.00098, '30': 1.00111, '35': 1.00126 },
            200:  { '-10': 1.00064, '0': 1.00076, '5': 1.00084, '10': 1.00092, '15': 1.00104, '20': 1.00116, '25': 1.00130, '30': 1.00148, '35': 1.00168 },
            250:  { '-10': 1.00080, '0': 1.00095, '5': 1.00105, '10': 1.00115, '15': 1.00130, '20': 1.00145, '25': 1.00163, '30': 1.00185, '35': 1.00210 },
            300:  { '-10': 1.00096, '0': 1.00114, '5': 1.00126, '10': 1.00138, '15': 1.00156, '20': 1.00174, '25': 1.00195, '30': 1.00221, '35': 1.00251 },
            350:  { '-10': 1.00112, '0': 1.00133, '5': 1.00147, '10': 1.00161, '15': 1.00182, '20': 1.00203, '25': 1.00228, '30': 1.00258, '35': 1.00293 },
            400:  { '-10': 1.00128, '0': 1.00152, '5': 1.00168, '10': 1.00184, '15': 1.00207, '20': 1.00231, '25': 1.00260, '30': 1.00295, '35': 1.00335 },
            450:  { '-10': 1.00144, '0': 1.00171, '5': 1.00189, '10': 1.00207, '15': 1.00233, '20': 1.00260, '25': 1.00292, '30': 1.00332, '35': 1.00377 },
            500:  { '-10': 1.00160, '0': 1.00190, '5': 1.00210, '10': 1.00230, '15': 1.00259, '20': 1.00289, '25': 1.00325, '30': 1.00368, '35': 1.00418 },
            600:  { '-10': 1.00192, '0': 1.00228, '5': 1.00252, '10': 1.00276, '15': 1.00311, '20': 1.00347, '25': 1.00390, '30': 1.00442, '35': 1.00502 },
            700:  { '-10': 1.00224, '0': 1.00266, '5': 1.00294, '10': 1.00322, '15': 1.00363, '20': 1.00405, '25': 1.00455, '30': 1.00515, '35': 1.00585 },
            800:  { '-10': 1.00256, '0': 1.00304, '5': 1.00336, '10': 1.00368, '15': 1.00414, '20': 1.00462, '25': 1.00519, '30': 1.00588, '35': 1.00668 },
            900:  { '-10': 1.00288, '0': 1.00342, '5': 1.00378, '10': 1.00414, '15': 1.00466, '20': 1.00520, '25': 1.00584, '30': 1.00662, '35': 1.00751 },
            1000: { '-10': 1.00320, '0': 1.00380, '5': 1.00420, '10': 1.00460, '15': 1.00518, '20': 1.00578, '25': 1.00649, '30': 1.00735, '35': 1.00835 },
            1100: { '-10': 1.00352, '0': 1.00418, '5': 1.00462, '10': 1.00506, '15': 1.00570, '20': 1.00636, '25': 1.00714, '30': 1.00808, '35': 1.00918 },
            1200: { '-10': 1.00384, '0': 1.00456, '5': 1.00504, '10': 1.00552, '15': 1.00622, '20': 1.00694, '25': 1.00779, '30': 1.00882, '35': 1.01001 }
        };

        this.loadCSV(csvPath);
    }

    loadCSV(csvPath) {
        return new Promise((resolve, reject) => {
            if (!fs.existsSync(csvPath)) {
                console.warn(`Archivo CSV no encontrado: ${csvPath}. Usando valores de referencia.`);
                this.dataLoaded = false;
                resolve();
                return;
            }

            fs.createReadStream(csvPath)
            .pipe(csv())
            .on('data', (row) => {
                this.api23Data.push(row);
            })
            .on('end', () => {
                console.log(`API 23E data loaded: ${this.api23Data.length} registros`);
                this.dataLoaded = true;
                resolve();
            })
            .on('error', (err) => {
                console.error('Error cargando CSV:', err);
                reject(err);
            });
        });
    }

    // ==========================================
    // UTILIDADES
    // ==========================================

    interpolate(x, x1, x2, y1, y2) {
        if (x1 === x2) return y1;
        return y1 + (x - x1) * (y2 - y1) / (x2 - x1);
    }

    fahrenheitToCelsius(f) {
        return (f - 32) * 5 / 9;
    }

    celsiusToFahrenheit(c) {
        return (c * 9 / 5) + 32;
    }

    psiToKpa(psi) {
        return psi * 6.89476;
    }

    kpaToPsi(kpa) {
        return kpa / 6.89476;
    }

    // ==========================================
    // PRESIÓN DE VAPOR (Pe)
    // ==========================================

    /**
     * Calcula la presión de vapor de equilibrio (Pe) a partir del RVP
     * Usando correlación de API MPMS 11.2.2 / GPA TP-15
     *
     * @param {number} rvp - Reid Vapor Pressure en PSI (a 100°F / 37.8°C)
     * @param {number} tempC - Temperatura de operación en °C
     * @returns {number} Presión de vapor en kPa absoluta
     */
    calculatePeFromRVP(rvp, tempC) {
        // El RVP se mide a 100°F (37.78°C)
        // Usamos la ecuación de Clausius-Clapeyron simplificada para extrapolar
        // ln(P2/P1) = (ΔHvap/R) × (1/T1 - 1/T2)

        // Convertir RVP de PSI a kPa
        const rvp_kPa = rvp * 6.89476;

        // Temperatura de referencia RVP (100°F = 37.78°C = 310.93K)
        const T_rvp_K = 310.93;

        // Temperatura de operación en Kelvin
        const T_op_K = tempC + 273.15;

        // Constante calibrada para LPG (ΔHvap/R)
        // Calibrada con datos de computador de flujo real
        // RVP=180 PSI, T=84.75°F → Pe≈1004 kPa, ΔP≈445 kPa → Cpl=1.0032
        const C = 2350;

        // Calcular Pe usando Clausius-Clapeyron
        const Pe_kPa = rvp_kPa * Math.exp(C * (1/T_rvp_K - 1/T_op_K));

        return Pe_kPa;
    }

    /**
     * Obtiene la presión de vapor de equilibrio (Pe) basada en densidad y temperatura
     * @param {number} density - Densidad en kg/m³
     * @param {number} tempC - Temperatura en °C
     * @returns {number} Presión de vapor en kPa absoluta
     */
    getVaporPressure(density, tempC) {
        // Seleccionar tabla según densidad más cercana
        const densities = Object.keys(this.vaporPressureTables).map(Number).sort((a, b) => a - b);

        let selectedDensity = densities[0];
        let minDiff = Math.abs(density - densities[0]);

        for (const d of densities) {
            const diff = Math.abs(density - d);
            if (diff < minDiff) {
                minDiff = diff;
                selectedDensity = d;
            }
        }

        const table = this.vaporPressureTables[selectedDensity.toString()];
        const temps = Object.keys(table).map(Number).sort((a, b) => a - b);

        // Encontrar temperaturas para interpolar
        let lowTemp = temps[0];
        let highTemp = temps[temps.length - 1];

        if (tempC <= temps[0]) return table[temps[0]];
        if (tempC >= temps[temps.length - 1]) return table[temps[temps.length - 1]];

        for (let i = 0; i < temps.length - 1; i++) {
            if (tempC >= temps[i] && tempC <= temps[i + 1]) {
                lowTemp = temps[i];
                highTemp = temps[i + 1];
                break;
            }
        }

        return this.interpolate(tempC, lowTemp, highTemp, table[lowTemp], table[highTemp]);
    }

    // ==========================================
    // FACTOR CTL (Corrección por Temperatura)
    // ==========================================

    /**
     * Calcula el factor de corrección por temperatura (Ctl) para LPG
     * Basado en coeficientes de expansión térmica de API MPMS 11.2.4
     * @param {number} density - Densidad observada en kg/m³
     * @param {number} tempC - Temperatura en °C
     * @returns {number} Factor Ctl
     */
    calculateCtl(density, tempC) {
        // Coeficiente de expansión térmica para LPG
        // Varía según la densidad - valores aproximados de API
        // Para propano (~500-510 kg/m³): α ≈ 0.00162 /°C
        // Para butano (~580 kg/m³): α ≈ 0.00124 /°C

        let alpha;
        if (density < 520) {
            // Propano - coeficiente ajustado según datos reales
            // Para densidad ~500-510 kg/m³
            alpha = 0.00314;
        } else if (density < 560) {
            // Mezcla LPG
            alpha = 0.00260;
        } else {
            // Butano (~580 kg/m³)
            alpha = 0.00210;
        }

        const referenceTemp = 15; // °C (estándar API)
        const deltaT = tempC - referenceTemp;

        // Ctl = exp(-α × ΔT × (1 + 0.8 × α × ΔT))
        // Fórmula simplificada: Ctl ≈ 1 / (1 + α × ΔT)
        const Ctl = 1 / (1 + alpha * deltaT);

        return Ctl;
    }

    // ==========================================
    // FACTOR CPL (Corrección por Presión)
    // ==========================================

    /**
     * Calcula el factor de corrección por presión (Cpl) usando tablas API 11.2.2M
     * @param {number} deltaP_kPa - Presión diferencial (Pm - Pe) en kPa
     * @param {number} tempC - Temperatura en °C
     * @returns {number} Factor Cpl
     */
    calculateCpl(deltaP_kPa, tempC) {
        if (deltaP_kPa <= 0) {
            console.warn(`Advertencia: ΔP = ${deltaP_kPa.toFixed(1)} kPa. Presión insuficiente.`);
            return 1.0;
        }

        // Encontrar ΔP más cercano en la tabla
        const deltaPs = Object.keys(this.cplTable).map(Number).sort((a, b) => a - b);

        let lowDeltaP = deltaPs[0];
        let highDeltaP = deltaPs[deltaPs.length - 1];

        // Si está fuera de rango
        if (deltaP_kPa <= deltaPs[0]) {
            // Interpolar hacia 0
            const cplAtLowest = this.getCplFromTable(deltaPs[0], tempC);
            return this.interpolate(deltaP_kPa, 0, deltaPs[0], 1.0, cplAtLowest);
        }
        if (deltaP_kPa >= deltaPs[deltaPs.length - 1]) {
            // Extrapolar (usar el último valor)
            return this.getCplFromTable(deltaPs[deltaPs.length - 1], tempC);
        }

        for (let i = 0; i < deltaPs.length - 1; i++) {
            if (deltaP_kPa >= deltaPs[i] && deltaP_kPa <= deltaPs[i + 1]) {
                lowDeltaP = deltaPs[i];
                highDeltaP = deltaPs[i + 1];
                break;
            }
        }

        const cplLow = this.getCplFromTable(lowDeltaP, tempC);
        const cplHigh = this.getCplFromTable(highDeltaP, tempC);

        return this.interpolate(deltaP_kPa, lowDeltaP, highDeltaP, cplLow, cplHigh);
    }

    /**
     * Obtiene Cpl de la tabla para un ΔP y temperatura específicos
     */
    getCplFromTable(deltaP, tempC) {
        const row = this.cplTable[deltaP];
        if (!row) return 1.0;

        const temps = Object.keys(row).map(Number).sort((a, b) => a - b);

        if (tempC <= temps[0]) return row[temps[0]];
        if (tempC >= temps[temps.length - 1]) return row[temps[temps.length - 1]];

        let lowTemp = temps[0];
        let highTemp = temps[temps.length - 1];

        for (let i = 0; i < temps.length - 1; i++) {
            if (tempC >= temps[i] && tempC <= temps[i + 1]) {
                lowTemp = temps[i];
                highTemp = temps[i + 1];
                break;
            }
        }

        return this.interpolate(tempC, lowTemp, highTemp, row[lowTemp], row[highTemp]);
    }

    // ==========================================
    // MÉTODO PRINCIPAL
    // ==========================================

    /**
     * Convierte flujo másico (kg/h) a flujo volumétrico estándar (a 60°F y Pe)
     *
     * @param {number} massFlowRate - Flujo másico en kg/h
     * @param {string} fluid - Tipo de fluido: 'butano', 'propano', 'lpg'
     * @param {number} observedTempF - Temperatura observada en °F
     * @param {number} observedPressurePSIG - Presión observada en PSIG (gauge)
     * @param {number} observedDensity - Densidad observada en kg/m³ (del Coriolis)
     * @param {number} [rvp] - Reid Vapor Pressure en PSI (opcional, del laboratorio)
     * @param {number} [mFactor=1.0] - Factor de corrección del medidor (opcional)
     * @returns {object} Flujos volumétricos y factores de corrección
     */
    convertMassToStandardVolume(massFlowRate, fluid, observedTempF, observedPressurePSIG, observedDensity, rvp = null, mFactor = 1.0) {
        const fluidKey = fluid.toLowerCase();

        // Convertir temperatura a °C
        const tempC = this.fahrenheitToCelsius(observedTempF);

        // Convertir presión a kPa absoluta
        const Pm_kPa = this.psiToKpa(observedPressurePSIG) + 101.325; // Gauge a absoluta

        // 1. Calcular presión de vapor de equilibrio (Pe)
        let Pe_kPa;
        let peSource;

        if (rvp !== null && rvp > 0) {
            // Usar RVP del laboratorio
            Pe_kPa = this.calculatePeFromRVP(rvp, tempC);
            peSource = 'RVP';
        } else {
            // Usar tablas estimadas
            Pe_kPa = this.getVaporPressure(observedDensity, tempC);
            peSource = 'Tabla';
        }

        // 2. Calcular presión diferencial (ΔP)
        const deltaP_kPa = Pm_kPa - Pe_kPa;

        // 3. Calcular factor de corrección por temperatura (Ctl)
        const Ctl = this.calculateCtl(observedDensity, tempC);

        // 4. Calcular factor de corrección por presión (Cpl)
        const Cpl = this.calculateCpl(deltaP_kPa, tempC);

        // 5. Factor combinado (CCF o CTPL)
        const CTPL = Ctl * Cpl;

        // 6. Densidad estándar (corregida a 60°F)
        const standardDensity = observedDensity / Ctl;

        // 7. Flujo volumétrico indicado (IV) - Masa / Densidad Observada
        const volumeFlowObserved_m3h = massFlowRate / observedDensity;
        const volumeFlowIndicado_gal = volumeFlowObserved_m3h * 264.172;

        // 8. Flujo volumétrico bruto (GUV) - Indicado × M-Factor
        const volumeFlowGross_gal = volumeFlowIndicado_gal * mFactor;
        const volumeFlowGross_m3h = volumeFlowGross_gal / 264.172;

        // 9. Flujo volumétrico estándar (GSV) - Bruto × CCF
        const volumeFlowStandard_gal = volumeFlowGross_gal * CTPL;
        const volumeFlowStandard_m3h = volumeFlowStandard_gal / 264.172;

        return {
            // Flujo volumétrico estándar (GSV) a 60°F
            m3_h: volumeFlowStandard_m3h,
            m3_d: volumeFlowStandard_m3h * 24,
            lt_h: volumeFlowStandard_m3h * 1000,
            lt_min: volumeFlowStandard_m3h * 1000 / 60,
            bbl_h: volumeFlowStandard_m3h * 6.28981,
            bbl_d: volumeFlowStandard_m3h * 6.28981 * 24,
            gal_h: volumeFlowStandard_gal,
            gal_min: volumeFlowStandard_gal / 60,
            ft3_h: volumeFlowStandard_m3h * 35.3147,

            // Volúmenes según nomenclatura del computador de flujo
            volumes: {
                indicated_USgal: volumeFlowIndicado_gal,    // IV - Volumen indicado (Masa/DensObs)
                gross_USgal: volumeFlowGross_gal,           // GUV - Volumen bruto (IV × M-Factor)
                standard_USgal: volumeFlowStandard_gal,     // GSV - Volumen estándar (GUV × CCF)
                mass_kg: massFlowRate
            },

            // Flujo volumétrico bruto (GUV)
            gross: {
                m3_h: volumeFlowGross_m3h,
                lt_h: volumeFlowGross_m3h * 1000,
                gal_h: volumeFlowGross_gal
            },

            // Flujo a condiciones observadas (IV)
            observed: {
                m3_h: volumeFlowObserved_m3h,
                lt_h: volumeFlowObserved_m3h * 1000,
                gal_h: volumeFlowIndicado_gal
            },

            // Datos de entrada
            input: {
                massFlowRate_kg_h: massFlowRate,
                fluid: fluidKey,
                observedTemp_F: observedTempF,
                observedTemp_C: tempC,
                observedPressure_PSIG: observedPressurePSIG,
                observedPressure_kPa_abs: Pm_kPa,
                observedDensity_kg_m3: observedDensity,
                observedDensity_SG: observedDensity / this.WATER_DENSITY_60F,
                mFactor: mFactor,
                rvp: rvp
            },

            // Factores de corrección
            corrections: {
                Ctl: Ctl,
                Cpl: Cpl,
                CCF: CTPL,              // Combined Correction Factor (= CTPL)
                CTPL: CTPL,
                mFactor: mFactor,
                Pe_kPa: Pe_kPa,
                Pe_PSIA: this.kpaToPsi(Pe_kPa),
                Pe_PSIG: this.kpaToPsi(Pe_kPa) - 14.696,  // Pe en gauge
                Pe_source: peSource,
                deltaP_kPa: deltaP_kPa,
                deltaP_PSI: this.kpaToPsi(deltaP_kPa),
                standardDensity_kg_m3: standardDensity,
                standardDensity_SG: standardDensity / this.WATER_DENSITY_60F
            }
        };
    }

    /**
     * Método para Node-RED
     */
    processNodeRedMessage(msg) {
        const {
            massFlow,
            fluid,
            observedTempF,
            observedPressurePSI,
            observedDensity,
            rvp,                   // Reid Vapor Pressure en PSI (opcional)
            mFactor                // Factor de corrección del medidor (opcional)
        } = msg.payload;

        if (massFlow === undefined) throw new Error('Parámetro requerido: massFlow (kg/h)');
        if (!fluid) throw new Error('Parámetro requerido: fluid (butano/propano/lpg)');
        if (observedTempF === undefined) throw new Error('Parámetro requerido: observedTempF (°F)');
        if (observedPressurePSI === undefined) throw new Error('Parámetro requerido: observedPressurePSI (PSIG)');
        if (observedDensity === undefined) throw new Error('Parámetro requerido: observedDensity (kg/m³)');

        const result = this.convertMassToStandardVolume(
            massFlow,
            fluid,
            observedTempF,
            observedPressurePSI,
            observedDensity,
            rvp || null,
            mFactor || 1.0
        );

        return { ...msg, payload: result };
    }
}

module.exports = CalcularFlujoEstandar;


// ============================================
// EJEMPLO DE USO - Comparación con computador de flujo
// ============================================
if (require.main === module) {
    const calc = new CalcularFlujoEstandar();

    setTimeout(() => {
        console.log('\n' + '='.repeat(70));
        console.log('COMPARACIÓN CON COMPUTADOR DE FLUJO');
        console.log('='.repeat(70));

        // ========================================
        // EJEMPLO 1: Sin RVP (usando tablas)
        // ========================================
        console.log('\n--- EJEMPLO 1: SIN RVP (usando tablas estimadas) ---\n');

        const result1 = calc.convertMassToStandardVolume(
            18580.72,    // kg/h
            'propano',
            84.75,       // °F
            195.54,      // PSIG
            502.4        // kg/m³
            // Sin RVP
        );

        console.log('ENTRADA:');
        console.log(`  Masa:            ${result1.input.massFlowRate_kg_h} kg/h`);
        console.log(`  Temperatura:     ${result1.input.observedTemp_F} °F (${result1.input.observedTemp_C.toFixed(2)} °C)`);
        console.log(`  Presión:         ${result1.input.observedPressure_PSIG} PSIG`);
        console.log(`  Densidad:        ${result1.input.observedDensity_kg_m3} kg/m³`);
        console.log(`  RVP:             No proporcionado (usando tabla)`);

        console.log('\nFACTORES:');
        console.log(`  Ctl:     ${result1.corrections.Ctl.toFixed(4)}`);
        console.log(`  Cpl:     ${result1.corrections.Cpl.toFixed(4)}`);
        console.log(`  CCF:     ${result1.corrections.CCF.toFixed(4)}`);
        console.log(`  Pe:      ${result1.corrections.Pe_kPa.toFixed(1)} kPa (${result1.corrections.Pe_PSIA.toFixed(1)} PSIA) [${result1.corrections.Pe_source}]`);
        console.log(`  ΔP:      ${result1.corrections.deltaP_kPa.toFixed(1)} kPa`);

        // ========================================
        // EJEMPLO 2: Con RVP del laboratorio
        // ========================================
        console.log('\n' + '-'.repeat(70));
        console.log('\n--- EJEMPLO 2: CON RVP DEL LABORATORIO ---\n');

        // Supongamos que el laboratorio reportó un RVP de 180 PSI
        const rvpLab = 180; // PSI @ 100°F

        const result2 = calc.convertMassToStandardVolume(
            18580.72,    // kg/h
            'propano',
            84.75,       // °F
            195.54,      // PSIG
            502.4,       // kg/m³
            rvpLab       // RVP del laboratorio
        );

        console.log('ENTRADA:');
        console.log(`  Masa:            ${result2.input.massFlowRate_kg_h} kg/h`);
        console.log(`  Temperatura:     ${result2.input.observedTemp_F} °F (${result2.input.observedTemp_C.toFixed(2)} °C)`);
        console.log(`  Presión:         ${result2.input.observedPressure_PSIG} PSIG`);
        console.log(`  Densidad:        ${result2.input.observedDensity_kg_m3} kg/m³`);
        console.log(`  RVP:             ${rvpLab} PSI @ 100°F (del laboratorio)`);

        console.log('\nFACTORES:');
        console.log(`  Ctl:     ${result2.corrections.Ctl.toFixed(4)}`);
        console.log(`  Cpl:     ${result2.corrections.Cpl.toFixed(4)}`);
        console.log(`  CCF:     ${result2.corrections.CCF.toFixed(4)}`);
        console.log(`  Pe:      ${result2.corrections.Pe_kPa.toFixed(1)} kPa (${result2.corrections.Pe_PSIA.toFixed(1)} PSIA) [${result2.corrections.Pe_source}]`);
        console.log(`  ΔP:      ${result2.corrections.deltaP_kPa.toFixed(1)} kPa`);

        // ========================================
        // COMPARACIÓN
        // ========================================
        console.log('\n' + '-'.repeat(70));
        console.log('\n--- COMPARACIÓN CON COMPUTADOR DE FLUJO ---\n');
        console.log('                    Sin RVP    Con RVP    Computador');
        console.log(`  Ctl:              ${result1.corrections.Ctl.toFixed(4)}     ${result2.corrections.Ctl.toFixed(4)}     0.9570`);
        console.log(`  Cpl:              ${result1.corrections.Cpl.toFixed(4)}     ${result2.corrections.Cpl.toFixed(4)}     1.0032`);
        console.log(`  CCF:              ${result1.corrections.CCF.toFixed(4)}     ${result2.corrections.CCF.toFixed(4)}     0.9601`);
        console.log(`  Pe (PSIA):        ${result1.corrections.Pe_PSIA.toFixed(1)}      ${result2.corrections.Pe_PSIA.toFixed(1)}       ???`);

        // ========================================
        // BUSCAR RVP QUE COINCIDA CON COMPUTADOR
        // ========================================
        console.log('\n' + '-'.repeat(70));
        console.log('\n--- BÚSQUEDA DE RVP PARA COINCIDIR CON Cpl=1.0032 ---\n');

        // Iterar para encontrar el RVP que produce Cpl ≈ 1.0032
        for (let testRvp = 100; testRvp <= 220; testRvp += 10) {
            const testResult = calc.convertMassToStandardVolume(
                18580.72, 'propano', 84.75, 195.54, 502.4, testRvp
            );
            console.log(`  RVP=${testRvp} PSI → Pe=${testResult.corrections.Pe_PSIA.toFixed(1)} PSIA, ΔP=${testResult.corrections.deltaP_kPa.toFixed(0)} kPa, Cpl=${testResult.corrections.Cpl.toFixed(4)}`);
        }

    }, 500);
}
